File: top150/ArrayString/q134.java
package top150.ArrayString;

public class q134 {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int n = gas.length;
        int overflow = 0;
        int totaloverflow = 0;
        int start = 0;

        for(int i = 0; i < n; i++) {
            overflow += gas[i]-cost[i];
            totaloverflow += gas[i]-cost[i];

            if(overflow < 0) {
                start = i+1;
                overflow = 0;
            }
        }

        return totaloverflow >= 0 ? start: -1;
    }
}

File: top150/_01_ArrayString/q134.py
class Solution:
    # T: O(N), S: O(1)
    def canCompleteCircuit(self, gas, cost):
        n = len(gas)
        cur_surplus = 0
        total_surplus = 0
        start = 0
        for i in range(n):
            cur_surplus += gas[i] - cost[i]
            total_surplus += gas[i] - cost[i]
            if cur_surplus < 0:
                cur_surplus = 0
                start = i+1
        return -1 if total_surplus < 0 else start
